axlearn/common/attention.py

```diff
@@ -1808,11 +1819,15 @@ def _compute_attention(
             k_proj: [batch_size, source_length, num_heads, per_head_dim].
             v_proj: [batch_size, source_length, num_heads, per_head_dim].
             attention_logit_biases: See ``On attention logit biases`` in the file comments.
+            segment_ids: See ``segment_ids`` in the file comments.
 
         Returns:
             The context of shape [batch_size, target_length, num_heads, per_head_dim],
             and probs of shape [batch, num_heads, target_length, source_length].
         """
+        # Ignore segment_ids. attention_logit_biases should be used instead.
+        del segment_ids
```

Comment:
Rather than silently ignoring, should we warn if segment_ids is not None?

Reply:
We can also consider another behavior when `segment_id` is provided and the attention doesn't have native `segment_id` support: we can construct logit biases from `segment_id` and apply this additional `bias` to `attention_logit_biases`, regardless whether  `attention_logit_biases` is None or not.

The down side of this approach is that user cannot control whether to ignore `segment_id` in attention computation. Previously, this can be done by setting `attention_mask` layer config to `None`. But the new behavior might be more preferable, because it is consistent with the behavior of `FlashAttention`.

Reply:
Maybe we can raise a ValueError if segmend_ids are provided but `attention_logit_biases is None`?

Reply:
Sounds good - done!

Reply:
Sorry, I realized I didn't hit "submit review". Another option would have been to pass None for segment_ids at the caller if it has already been merged into the attention logit biases. At the moment we seem to be making this assumption implicitly at `MultiheadAttention`, but the logic only holds if callers follow `Decoder`.

Reply:
Agreed - I set `segment_ids = None` and merge it to `attention_logit_biases` at the caller when `attention_logit_biases` is set. I think this is cleaner.